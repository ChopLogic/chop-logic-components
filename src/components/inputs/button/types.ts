import { MouseEventHandler } from 'react';
import { CommonInputProps } from 'utils/types';

import { ChopLogicIconName } from 'components/misc/icon/Icon';

export type ChopLogicButtonView = '3D' | 'flat' | 'icon';

export type ChopLogicButtonProps = Partial<CommonInputProps> & {
  onClick?: MouseEventHandler<HTMLButtonElement>;
  view?: ChopLogicButtonView;
  icon?: ChopLogicIconName;
  text?: string;
  extended?: boolean;
  type?: 'button' | 'reset' | 'submit';
};
