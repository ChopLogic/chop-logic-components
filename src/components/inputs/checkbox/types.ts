export type ChopLogicCheckboxProps = React.InputHTMLAttributes<HTMLInputElement> & {
  id: string;
  name: string;
  label: string;
  isLabelHidden?: boolean;
};
