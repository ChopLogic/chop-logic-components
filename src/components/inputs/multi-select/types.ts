export type MultiSelectValue = {
  id: string;
  label: string;
  selected: boolean;
};

export type ChopLogicMultiSelectProps = React.SelectHTMLAttributes<HTMLSelectElement> & {
  id: string;
  name: string;
  label: string;
  values: MultiSelectValue[];
  onSelectChange?: (values?: MultiSelectValue[]) => void;
  placeholder?: string;
};

export type MultiSelectDropdownProps = {
  values: MultiSelectValue[];
  isOpened: boolean;
  dropdownId: string;
  onClose: () => void;
  onSelect: (id: string) => void;
};
