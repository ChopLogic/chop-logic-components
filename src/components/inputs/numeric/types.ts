export type NumericValidationFunction = (input?: number) => boolean;

export type ChopLogicNumericInputProps = React.InputHTMLAttributes<HTMLInputElement> & {
  name: string;
  label: string;
  errorMessage?: string;
  validator?: NumericValidationFunction;
  onClear?: () => void;
  hasSpinButtons?: boolean;
  onSpinButtonClick?: (value?: number) => void;
};
