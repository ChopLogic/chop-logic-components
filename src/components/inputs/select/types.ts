export type ChopLogicSelectProps = React.SelectHTMLAttributes<HTMLSelectElement> & {
  id: string;
  name: string;
  label: string;
  values: SelectValue[];
  onSelectChange?: (value?: SelectValue) => void;
  placeholder?: string;
};

export type SelectValue = {
  id: string;
  label: string;
};
