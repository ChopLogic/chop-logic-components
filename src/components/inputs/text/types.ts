export type ChopLogicTextInputProps = React.InputHTMLAttributes<HTMLInputElement> & {
  id: string;
  name: string;
  label: string;
  valid?: boolean;
  errorMessage?: string;
};
