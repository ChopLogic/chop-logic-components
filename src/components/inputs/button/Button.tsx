import React from 'react';
import { ThemeProvider } from 'styled-components';

import { ChopLogicIcon } from '@/elements';
import { ChopLogicButtonProps } from '@/types';
import { getChopLogicTheme } from '@/utils';

import { Styled3DButton, StyledFlatButton, StyledIconButton } from './Button.styled';

const ChopLogicButton: React.FC<ChopLogicButtonProps> = ({
  onClick,
  text,
  icon,
  label,
  disabled = false,
  extended = false,
  type = 'button',
  view = '3D',
  theme,
  ...rest
}) => {
  const themeValues = getChopLogicTheme(theme);

  switch (view) {
    case '3D':
      return (
        <ThemeProvider theme={themeValues}>
          <Styled3DButton $extended={extended} aria-label={label} type={type} onClick={onClick} disabled={disabled} {...rest}>
            <span className='chop-logic-button_shadow'></span>
            <span className='chop-logic-button_edge'></span>
            <span className='chop-logic-button_front'>
              <ChopLogicIcon name={icon} />
              {text}
            </span>
          </Styled3DButton>
        </ThemeProvider>
      );
    case 'flat':
      return (
        <ThemeProvider theme={themeValues}>
          <StyledFlatButton $extended={extended} aria-label={label} type={type} onClick={onClick} disabled={disabled} {...rest}>
            <ChopLogicIcon name={icon} />
            {text}
          </StyledFlatButton>
        </ThemeProvider>
      );
    case 'icon':
      return (
        <ThemeProvider theme={themeValues}>
          <StyledIconButton aria-label={label} type={type} onClick={onClick} disabled={disabled} {...rest}>
            <ChopLogicIcon name={icon} />
          </StyledIconButton>
        </ThemeProvider>
      );
    default:
      return null;
  }
};

export default ChopLogicButton;
