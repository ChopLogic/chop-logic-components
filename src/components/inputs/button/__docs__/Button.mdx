import { Canvas, Meta } from '@storybook/blocks';
import * as ButtonStories from './Button.stories';

<Meta of={ButtonStories} title='Button' />

# ChopLogicButton

Styled button component with different view modes and a customizable icon.

### Default View

<Canvas of={ButtonStories.DefaultButton} />

### Usage

```tsx
import { ChopLogicButton, ChopLogicIconName } from 'chop-logic-components';
import { createCopy } from 'utils/copy';

const Example = () => {
  return <ChopLogicButton text='Copy' onClick={createCopy} icon={ChopLogicIconName.Copy} />;
};

export default Example;
```

### Props

| Prop          | Type                                  | Description                                                                                                                                                                              | Default     | Required |
|---------------|---------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------|----------|
| **text**      | `string`                              | A string that represents the text content of the button.                                                                                                                                 | `undefined` | No       |
| **view**      | `ChopLogicButtonView`                 | An appearance of the button.                                                                                                                                                             | `undefined` | No       |
| **onClick**   | ` MouseEventHandler`                  | A function that is called when the button is clicked. It receives a MouseEventHandler for handling the click event on the button element. It could also be a callback without arguments. | `undefined` | No       |
| **disabled**  | `boolean`                             | A boolean indicating whether the button should be disabled or not. When disabled, the button cannot be clicked or interacted with.                                                       | `undefined` | No       |
| **type**      | `'button'` or `'reset'` or `'submit'` | The default behavior of the button.                                                                                                                                                      | `'button'`  | No       |
| **icon**      | `ChopLogicIconName`                   | An svg icon from ChopLogicIcon set, which could be inserted before the button text.                                                                                                      | `undefined` | No       |
| **id**        | `string`                              | The id global attribute defines an identifier (ID) which must be unique in the whole document.                                                                                           | `undefined` | No       |
| **className** | `string`                              | The class global attribute is a list of the classes of the element.                                                                                                                      | `undefined` | No       |
| **style**     | `CSSProperties`                       | The style global attribute contains CSS styling declarations to be applied to the element.                                                                                               | `undefined` | No       |
| **tabIndex**  | `number`                              | The tabindex global attribute allows developers to make HTML elements focusable.                                                                                                         | `undefined` | No       |
| **title**     | `string`                              | The title global attribute contains text representing advisory information related to the element it belongs to.                                                                         | `undefined` | No       |
| **theme**     | `'light'` or `'dark'`                 | The component's color theme.                                                                                                                                                             | `undefined` | No       |

### Types

```ts
export type ChopLogicButtonProps = Partial<CommonInputProps> & {
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  view?: ChopLogicButtonView;
  icon?: ChopLogicIconName;
  text?: string;
  extended?: boolean;
  type?: 'button' | 'reset' | 'submit';
};

export type ChopLogicButtonView = '3D' | 'flat' | 'icon';

export type CommonInputProps = CommonComponentProps & {
  label: string;
  name: string;
  disabled?: boolean;
  required?: boolean;
};

export type CommonComponentProps = {
  id?: string;
  className?: string;
  style?: React.CSSProperties;
  tabIndex?: number;
  title?: string;
  theme?: 'dark' | 'light';
};
```
