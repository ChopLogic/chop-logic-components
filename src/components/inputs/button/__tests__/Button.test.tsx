import { render, screen } from '@testing-library/react';
import { PointerEventsCheckLevel, userEvent } from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import { CLIcon } from 'components/misc/icon/types';

import Button from '../Button';
import { ChopLogicButtonView } from '../types';

describe('ChopLogicButton', () => {
  const testProps = {
    id: 'test-button-id',
    text: 'Test button',
    icon: CLIcon.Copy,
    className: 'test-button-class',
    view: 'danger' as ChopLogicButtonView,
    disabled: false,
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<Button {...testProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should have the accessible role', () => {
    render(<Button {...testProps} />);
    expect(screen.getByRole('button')).toBeInTheDocument();
  });

  it('should not render the text if view is icon', () => {
    render(<Button {...testProps} view='icon' />);
    expect(screen.queryByRole(testProps.text)).not.toBeInTheDocument();
  });

  it('should call the callback function on click', async () => {
    const mockedCallback = vi.fn();
    render(<Button {...testProps} onClick={mockedCallback} />);
    await userEvent.click(screen.getByRole('button'));
    expect(mockedCallback).toHaveBeenCalledOnce();
  });

  it('should not call the callback function on click if the button is disabled', async () => {
    const mockedCallback = vi.fn();
    render(<Button {...testProps} onClick={mockedCallback} disabled={true} />);
    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
    await userEvent.click(button, {
      pointerEventsCheck: PointerEventsCheckLevel.Never,
    });
    expect(mockedCallback).not.toHaveBeenCalled();
  });

  it('should have focus on Tab press', () => {
    render(<Button {...testProps} />);
    const button = screen.getByRole('button');
    expect(button).not.toHaveFocus();
    button.focus();
  });

  it('should call the callback function on Enter press', async () => {
    const mockedCallback = vi.fn();
    render(<Button {...testProps} onClick={mockedCallback} />);
    const button = screen.getByRole('button');
    button.focus();
    await userEvent.keyboard('{Enter}');
    expect(mockedCallback).toHaveBeenCalledOnce();
  });
});
