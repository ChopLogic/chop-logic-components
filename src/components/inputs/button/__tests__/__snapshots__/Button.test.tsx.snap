// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChopLogicButton > should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-blHHSb kzfXet button-class"
    id="button-id"
    type="button"
  >
    <span
      class="chop-logic-button_shadow"
    />
    <span
      class="chop-logic-button_edge"
    />
    <span
      class="chop-logic-button_front"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M11 9 H20 A2 2 0 0 1 22 11 V20 A2 2 0 0 1 20 22 H11 A2 2 0 0 1 9 20 V11 A2 2 0 0 1 11 9 z"
        />
        <path
          d="M5 15H4a2 2 0 01-2-2V4a2 2 0 012-2h9a2 2 0 012 2v1"
        />
      </svg>
      Test button
    </span>
  </button>
</DocumentFragment>
`;
